ocaml/xapi/repository_helpers.ml

```diff
@@ -1361,15 +1324,139 @@ let prune_updateinfo_for_livepatches livepatches updateinfo =
   in
   {updateinfo with livepatches= lps}
 
-let do_with_device_models ~__context ~host f =
-  (* Call f with device models of all running HVM VMs on the host *)
-  Db.Host.get_resident_VMs ~__context ~self:host
-  |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
-  |> List.filter (fun (_, record) -> not record.API.vM_is_control_domain)
-  |> List.filter_map f
-  |> function
-  | [] ->
-      ()
-  | _ :: _ ->
-      let host' = Ref.string_of host in
-      raise Api_errors.(Server_error (cannot_restart_device_model, [host']))
+let do_with_host_pending_guidances ~op guidances =
+  List.iter
+    (fun g ->
+      match Guidance.to_pending_guidance g with
+      | ( Some `restart_toolstack
+        | Some `reboot_host
+        | Some `reboot_host_on_xen_livepatch_failure
+        | Some `reboot_host_on_kernel_livepatch_failure
+        | Some `reboot_host_on_livepatch_failure ) as g' ->
+          Option.iter op g'
+      | _ ->
+          ()
+    )
+    guidances
+
+let do_with_vm_pending_guidances ~op ~vm guidances =
+  List.iter
+    (fun g ->
+      match Guidance.to_pending_guidance g with
+      | Some `restart_device_model ->
+          op vm `restart_device_model
+      | Some `restart_vm ->
+          op vm `restart_vm
+      | _ ->
+          ()
+    )
+    guidances
+
+let merge_with_pending_guidances ~pending ~coming =
+  let open GuidanceSet in
+  let pending = of_list pending in
+  union pending (of_list coming) |> remove EvacuateHost |> fun unioned ->
```

Comment:
For mandatory guidances we will not filter guidance will partial order, we will only remove any duplicated ones. Then when will we need to remove some guidance from pending ones?
Is the `to_be_removed` guidances only would be `EvacuateHost`? But I think `EvacuateHost` will not exist in the legacy `host.pending_guidances` neither?

Would you please give an example when there would be `to_be_removed` returned here? Thank you!

Reply:
Indeed, `to_be_removed` will be empty always in the current implementation.  It is to keep an ability of removing something from the list.

Reply:
Understand, we can keep this logic, but would be better to add a comment to avoid any confusion later.